waiter/src/waiter/auth/authentication.clj

```diff
@@ -201,3 +228,110 @@
+        {:keys [expires-at]} auth-metadata
+        cookie-valid? (decoded-auth-valid? decoded-auth-cookie)
+        sanitized-expires-at (or (when cookie-valid? expires-at) 0)]
+    (log/info "waiter auth cookie parsed" {:auth-principal auth-principal :cookie-valid? cookie-valid? :expires-at expires-at})
+    (utils/attach-waiter-source (utils/clj->json-response {:expires-at sanitized-expires-at
+                                                           :principal auth-principal}))))
+
+(defn attach-authorization-headers
+  "Attaches authentication description headers into the response."
+  [{:keys [authorization/method authorization/principal authorization/user] :as response}]
+  (update response :headers
+          (fn [headers]
+            (cond-> headers
+              method (assoc "x-waiter-auth-method" (name method))
+              principal (assoc "x-waiter-auth-principal" (str principal))
+              user (assoc "x-waiter-auth-user" (str user))))))
+
+(defn remove-done-param-from-keep-alive-https-redirect
+  "Removes the done=true query parameter from the response when it is an https redirect to the keep-alive endpoint."
+  [{:keys [headers status] :as response}]
+  (let [{:strs [location]} headers
+        location-uri (when-not (str/blank? location)
+                       (URI. location))]
+    (cond-> response
+      (and location-uri
+           (contains? #{http-302-moved-temporarily http-307-temporary-redirect} status)
+           (= "https" (.getScheme location-uri))
+           (= auth-keep-alive-uri (.getPath location-uri)))
+      (assoc-in [:headers "location"]
+                (str/replace location auth-keep-alive-done-parameter auth-keep-alive-done-parameter-replacement)))))
+
+(defn process-auth-keep-alive-request
+  "Handler to eagerly trigger authentication workflow even if cookie has not yet expired.
+   This allows clients to pre-emptively refresh credentials before they expire.
+   Presence of done parameter is used to avoid infinite auth redirect loops and will return 204 No Content.
+   Invalid offset parameters result in a 400 error.
+   Missing offset, soon to expire cookie (based on offset) or invalid auth cookie will trigger the authentication workflow.
+   If cookie is expected to be live longer than offset value, return 204 No Content."
+  [password auth-handler {:keys [headers query-string] :as request}]
+  (let [{:strs [done offset]} (-> request ru/query-params-request :query-params)
+        offset-parsed (utils/parse-int offset)
+        decoded-auth-cookie (get-and-decode-auth-cookie-value headers password)
+        [auth-principal _ cookie-metadata] decoded-auth-cookie
+        current-epoch-time (tc/to-epoch (t/now))
+        {:keys [expires-at]} cookie-metadata
+        ;; handle legacy cookies which will not have this value set
+        expires-at (or expires-at current-epoch-time)
+        waiter-token (some-> request utils/request->host utils/authority->host)]
```

Comment:
This isn't a waiter token until it's been validated as such.

Reply:
Renamed variable, but it is treated as a Waiter token later on in the code.